File: baitaptuan3/Julia/bai3.jl
mutable struct Robot
    position::Vector{Int}  # [x, y]
    direction::String      # N, E, S, W
    history::Vector{String} # Lịch sử các lệnh

    Robot() = new([0, 0], "N", String[])
end

function move_forward(robot::Robot)
    if robot.direction == "N"
        robot.position[2] += 1
    elseif robot.direction == "E"
        robot.position[1] += 1
    elseif robot.direction == "S"
        robot.position[2] -= 1
    elseif robot.direction == "W"
        robot.position[1] -= 1
    end
    push!(robot.history, "forward")
    display_status(robot, "Đang tiến lên")
end

function move_backward(robot::Robot)
    if robot.direction == "N"
        robot.position[2] -= 1
    elseif robot.direction == "E"
        robot.position[1] -= 1
    elseif robot.direction == "S"
        robot.position[2] += 1
    elseif robot.direction == "W"
        robot.position[1] += 1
    end
    push!(robot.history, "backward")
    display_status(robot, "Đang lùi lại")
end

function turn_left(robot::Robot)
    dirs = ["N", "W", "S", "E"]
    idx = findfirst(==(robot.direction), dirs)
    robot.direction = dirs[mod(idx, 4) + 1]
    push!(robot.history, "left")
    display_status(robot, "Đang quay trái")
end

function turn_right(robot::Robot)
    dirs = ["N", "E", "S", "W"]
    idx = findfirst(==(robot.direction), dirs)
    robot.direction = dirs[mod(idx, 4) + 1]
    push!(robot.history, "right")
    display_status(robot, "Đang quay phải")
end

function undo(robot::Robot)
    if isempty(robot.history)
        println("Không có lệnh nào để hoàn tác.")
        return
    end

    last_command = pop!(robot.history)
    println("Hoàn tác lệnh: ", last_command)

    if last_command == "forward"
        move_backward(robot)
    elseif last_command == "backward"
        move_forward(robot)
    elseif last_command ==

File: baitaptuan3/C++/bai3.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <string>

using namespace std;

class Robot {
private:
    vector<int> position;  // [x, y]
    string direction;      // N, E, S, W
    stack<string> history; // Lịch sử các lệnh

public:
    Robot() : position({0, 0}), direction("N") {}

    void move_forward() {
        if (direction == "N") position[1]++;
        else if (direction == "E") position[0]++;
        else if (direction == "S") position[1]--;
        else if (direction == "W") position[0]--;
        history.push("forward");
        display_status("Đang tiến lên");
    }

    void move_backward() {
        if (direction == "N") position[1]--;
        else if (direction == "E") position[0]--;
        else if (direction == "S") position[1]++;
        else if (direction == "W") position[0]++;
        history.push("backward");
        display_status("Đang lùi lại");
    }

    void turn_left() {
        string dirs[] = {"N", "W", "S", "E"};
        int idx = (find(begin(dirs), end(dirs), direction) - begin(dirs) + 1) % 4;
        direction = dirs[idx];
        history.push("left");
        display_status("Đang quay trái");
    }

    void turn_right() {
        string dirs[] = {"N", "E", "S", "W"};
        int idx = (find(begin(dirs), end(dirs), direction) - begin(dirs) + 1) % 4;
        direction = dirs[idx];
        history.push("right");
        display_status("Đang quay phải");
    }

    void undo() {
        if (history.empty()) {
            cout << "Không có lệnh nào để hoàn tác." << endl;
            return;
        }

